python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -122,6 +122,31 @@ def _step_output_error_checked_user_event_sequence(
                     "must yield DynamicOutput, got Output."
                 )
 
+            # For any output associated with an asset, make sure that it has been yielded only after
+            # any dependency assets in the same step have been yielded. Without this chcek,
+            # downstream assets may be yielded before their dependencies, which disrupts the logical
+            # version computation.
+            asset_layer = step_context.pipeline_def.asset_layer
+            node_handle = step_context.solid_handle
+            asset_info = asset_layer.asset_info_for_output(node_handle, output_def.name)
+            if (
+                asset_info is not None
+                and asset_info.is_required
+                and asset_layer.has_assets_def_for_asset(asset_info.key)
+            ):
+                assets_def = asset_layer.assets_def_for_asset(asset_info.key)
+                if assets_def is not None:
+                    all_dep_keys = asset_layer.upstream_assets_for_asset(asset_info.key)
+                    step_local_asset_keys = assets_def.asset_keys
+                    step_local_dep_keys = all_dep_keys & step_local_asset_keys
```

Comment:
I think this logic breaks down for graph-backed assets, because not all of the asset keys in the definition may be produced in the same step.

might be misinterpreting, though, and I don't see an easy way to get the set of asset keys produced by a given step (rather than by a given node, where the node is the GraphDefinition).

Just for reference, if this is logic is incorrect, then https://sourcegraph.com/github.com/dagster-io/dagster/-/blob/python_modules/dagster/dagster/_core/execution/context/compute.py?L307&subtree=true is also incorrect, so I'm approving for now as this is not a net new issue, but I do think we should address the issue in a PR soon or make sure that it's not actually an issue.

cc @clairelin135 for your more seasoned opinion.

Reply:
Yep, this logic would incorrectly throw an error when assets are yielded from different steps. For example, this code snippet below would throw an error. I can also imagine edge cases where an output name is shared by a graph and by an op.
```
@graph(out={"out": GraphOut(), "out_plus_one": GraphOut()})
def my_graph():
    x = upstream()
    return {"out_plus_one": downstream(x), "out": x}


AssetsDefinition.from_graph(
    my_graph, internal_asset_deps={"out_plus_one": {AssetKey("out")}, "out": set()}
)
```
I think what we can do instead is determine the `step_local_asset_keys` by finding all the assets outputted from the step from the [`asset_info_by_node_output_handle` field on the asset layer](https://github.com/dagster-io/dagster/blob/f37fd6cb2844689eea8b406a53254d5d96720ac0/python_modules/dagster/dagster/_core/definitions/asset_layer.py#L503). Presumably any asset dependencies have already outputted in upstream steps and thus have materializations.

Reply:
After a long hiatus I've modified this PR to address the above feedback. But it looks like since I initially posted it, something has changed in `dagster-dbt`-- there is a test in there that is yielding outputs in non-topo-order. I don't know enough about `dagster-dbt` to quickly determine whether this is stemming from `dagster-dbt` itself or test design-- @OwenKephart mind having a look a the failing test?